File: 2023_8/baekjoon/17299.py
import sys

N = int(input())
arr = list(map(int, sys.stdin.readline().split()))
dict = {i: 0 for i in range(1, 1000001)}
for i in arr:
    dict[i] += 1

stack = [0]
result = [-1] * N

for i in range(1, N):
    v = arr[i]
    while stack and dict[arr[stack[-1]]] < dict[v]:
        result[stack.pop()] = v
    stack.append(i)
print(*result)

File: 2023_8/baekjoon/17299.rs
use std::io::{stdin, Read};

fn main() {
    let mut buf = String::new();
    stdin().read_to_string(&mut buf).unwrap();
    let mut buf = buf.split_ascii_whitespace().map(str::parse).flatten();
    let n = buf.next().unwrap();
    let mut freq = vec![0; 1_000_001];
    let arr: Vec<usize> = buf.take(n).inspect(|v| freq[*v] += 1).collect();
    let mut stack: Vec<usize> = vec![0];
    let mut result: Vec<i32> = vec![-1; n];

    for i in 0..n {
        while !stack.is_empty() && freq[arr[*stack.last().unwrap()]] < freq[arr[i]] {
            result[stack.pop().unwrap()] = arr[i] as i32;
        }
        stack.push(i);
    }
    println!(
        "{}",
        result
            .iter()
            .map(|v| v.to_string())
            .collect::<Vec<String>>()
            .join(" ")
    );
}
